src/main/kotlin/com/bardsoftware/backend/fleet/rmarkdown/Subscriber.kt

```diff
@@ -22,35 +22,100 @@ import com.google.cloud.pubsub.v1.Subscriber
 import com.google.pubsub.v1.PubsubMessage
 import com.google.pubsub.v1.SubscriptionName
 import com.xenomachina.argparser.ArgParser
+import org.apache.commons.io.FileUtils.readFileToString
+import java.io.ByteArrayInputStream
+import java.io.File
+import java.io.FileOutputStream
+import java.nio.charset.Charset
+import java.nio.file.Path
+import java.nio.file.Paths
 import java.util.concurrent.CompletableFuture
+import java.util.zip.ZipEntry
+import java.util.zip.ZipInputStream
 
 class SubscriberArgs(parser: ArgParser) {
     val subscriberName by parser.storing(
             "--sub",
             help = "subscription topic name")
+
+    val tasksDir by parser.storing(
+            "--tasks-dir",
+            help = "root of tasks content"
+    )
 }
 
 private val PROJECT_ID = ServiceOptions.getDefaultProjectId()
 
-internal class MessageReceiverExample(private val callback: (message: String, md5sum: String) -> Unit) : MessageReceiver {
+internal class TaskReceiver(private val tasksDir: Path,
+                            private val callback: (message: String, md5sum: String) -> Unit
+) : MessageReceiver {
     private val dockerProcessor = DockerProcessor()
+    private val BUFFER_SIZE = 4096
+
 
     override fun receiveMessage(message: PubsubMessage, consumer: AckReplyConsumer) {
         processMessage(message)
         consumer.ack()
     }
 
     fun processMessage(message: PubsubMessage) {
-        val messageContent = message.data.toStringUtf8()
-        val md5sum = this.dockerProcessor.getMd5Sum(messageContent)
+        val request = CompilerFleet.CompilerFleetRequest.parseFrom(message.data)
+
+        val taskId = request.taskId
+        val destination = this.tasksDir.resolve(taskId).resolve("files")
+        val zipByteInput = ByteArrayInputStream(request.zipBytes.toByteArray())
+        val zipStream = ZipInputStream(zipByteInput)
+        var entry: ZipEntry? = zipStream.nextEntry
+        val buffer = ByteArray(this.BUFFER_SIZE)
+
+        while (entry != null) {
+            val filename = entry.name
+            val newFile = destination.resolve(filename).toFile()
+
+            File(newFile.parent).mkdirs()
+            val fos = FileOutputStream(newFile)
+
+            var len: Int = zipStream.read(buffer)
+            while (len > 0) {
+                fos.write(buffer, 0, len)
+                len = zipStream.read(buffer)
+            }
+
+            fos.close()
```

Comment:
Please surround it into try with resources, .use, .with or whatever will automatically close the output stream should any exception occur. And do catch and process IOException e.g. by logging it.

Reply:
should I put here logging system such as Log4j?

Reply:
slf4j + logback is the choice. You can do it in a separate pull request though.

Reply:
I'll add logging in a separate PR